### ISSUE TITLE
[FR] Improve Spec.guess' guessing capabilities and remove hardcoded defaults

### ISSUE THREAD

**Issue by madskjeldgaard** [supercollider/supercollider#5516]
## Motivation
Currently, the JITLib gui stuff depends on Spec.guess when it cannot find a spec globally for the parameter in question. The result of this guess is used in building the gui and determining the value range of the slider. 

Currently, this is what the guess method looks like:

https://github.com/supercollider/supercollider/blob/de60e3405b417779bd35c19add1c6b2cca3bd194/SCClassLibrary/JITLib/GUI/extSoftSet.sc#L2-L12

If the parameter's value is a number, the guess method will return either one of two hardcoded and slightly arbitrary specs that will almost never result in a correct slider range and the result is that the NdefGui for example suffers from this down the line.

## Description of Proposed Feature
Can we use the global spec library to make the guessing mechanism smarter / more informed? There's also been talk of using fuzzy searching to perhaps find stuff in the global specs' keys to find something reasonable.

## Plan for Implementation

Sorry, I'm on a ferry and it's about to go into harbor....

**Comment by adcxyz** [comment]
Sorry I missed that question! 

Generally, the (undocumented) idea of guess was that users may write things 
without knowing good parameters ranges in advance.  
In that case, keep things going by making a first guess with something reasonable, 
and defining/tuning your ranges as you go. 
Assume the initial value is somewhere in the middle of the range, 
So the first pass for initial value 0 is that you might want to go both positive and negative, 
so it would work for pan, balance, pos of PanAz etc (+-1); make it +-2 so there is more range to explore. 
If nonzero, assume an exponential range (as many parameters are perceptually exponential), 
and allow a wide exponential range of x / 20 to x * 20. 

From there, you refine that param specs generally with 
```Spec.add(\param, [...]);```
or specifically for the Ndef with addSpec, in JITLibExtensions. 
```Ndef(\x).addSpec(\param, [...]);```

That said, Spec.guess could just be a user-definable function that handles 
the guessing conditions as desired, something like:
```
Spec.guessFunc = { ... };

Spec { 
	...
	*guess { |key, value|
		// 
		guessFunc !? { ^guessFunc.value(key, value) };
		// and keep default for compatibility:
		if (value.isKindOf(SimpleNumber).not) { ^nil };

		^if (value.abs > 0) {
			ControlSpec(value/20, value*20, \exp, 0, value, \guess);
		} {
			ControlSpec(-2, 2, \lin, 0, 0, \guess);
		};
	}
}
```
Does that make sense?

**Comment by telephon** [comment]
I suppose both could be changed together. If we call `this.guessFunc` one may check `respondsTo(\guessFunc)` to be backward compatible.

**Comment by redFrik** [comment]
sorry, i wasn't clear.
i was more wondering if, while having JITLibExtensions installed together with the guessFunc implementation suggested above, i'd need to perform checks for isFilterRole and/or isKindOf(Array) inside any custom guessFunc i provide.
or can a custom guessFunc safely ignore those and be really simple.  e.g. just return [-1, 1].asSpec?
but then what happens if i give a parameter that _is_ true for isFilterRole and need(?) to be [0, 1].asSpec?  e.g. Ndef(\f, {|mix5=3| DC.ar(0)}).gui.  this will currently run through JITLibExtensions guess variant, but later my custom guessFunc would override this.
i have a hunch that something here isn't completely watertight.

**Comment by telephon** [comment]
Ok, yes, I'm also not sure.
I am adding the specs directly to the control, like in:

```supercollider
Ndef(\y, { |in| \xamp.kr(1, spec:[1, 3, \exp].asSpec) }).gui;
```

**Comment by capital-G** [comment]
Maybe OT but I personally like the current guess implementation - it does work most of the time rather good but is indeed somehow annoying if it guesses wrong.

The main problem seems to me that if I hit a barrier in the GUI and want to extend the slider beyond its current barrier - it would be great if - in JIT tradition - I could change those barriers (e.g. through a button).
E.g. I discover that my `\customParam` starts to sounds great above 2.0, but there is a GUI limit - instead of closing and reloading the GUI window to extend the range, it would be great to have a button to change the barriers of the GUI (e.g. make current value "center" with some heuristics on the range it has?)

Therefore - it seems to me more important that such a "guessing" should be dynamic and updateable and have an reaction in the GUI (or elsewhere) b/c I may not know in advance which parameter range is interesting and want to modify those barrier while its running.

**Comment by telephon** [comment]
Good idea to use the ControlSpec itself for further guessing or extending. Some methods may be useful as well, where you can say, e.g. `spec.extend(min:-1)` or `spec.extendByFactor(2)`
